### PR TITLE
Convert CertificatePinner to Kotlin

### PR DESCRIPTION
**Pull request by swankjesse**

### CODE REVIEW

File: okhttp/src/main/java/okhttp3/CertificatePinner.java [hunk 1]
```diff
@@ -13,333 +13,324 @@
...
-          if (sha1 == null) sha1 = sha1(x509Certificate);
-          if (pin.hash.equals(sha1)) return; // Success!
-        } else {
-          throw new AssertionError("unsupported hashAlgorithm: " + pin.hashAlgorithm);
+      var sha1: ByteString? = null
+      var sha256: ByteString? = null
+
+      for (pin in pins) {
+        when (pin.hashAlgorithm) {
+           "sha256/" -> {
```
Comment (Egorand) [inline okhttp/src/main/java/okhttp3/CertificatePinner.java]: nit: wrong indentation

File: okhttp/src/main/java/okhttp3/CertificatePinner.kt [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+        append("\n    ")
+        append(pin)
+      }
+    }
+    throw SSLPeerUnverifiedException(message)
+  }
+
+  @Deprecated("replaced with {@link #check(String, List)}. ")
+  @Throws(SSLPeerUnverifiedException::class)
+  fun check(hostname: String, vararg peerCertificates: Certificate) {
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/CertificatePinner.kt]: Inline function?

File: okhttp/src/main/java/okhttp3/CertificatePinner.kt [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+      append(":")
+      for (pin in pins) {
+        append("\n    ")
+        append(pin)
+      }
+    }
+    throw SSLPeerUnverifiedException(message)
+  }
+
+  @Deprecated("replaced with {@link #check(String, List)}. ")
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/CertificatePinner.kt]: Migrate to `ReplaceWith`

File: okhttp/src/test/java/okhttp3/CertificatePinnerKotlinTest.kt [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    assertThat(certificatePinner.findMatchingPins("a.b.example.com")).isEmpty()
+  }
+
+  @Test fun successfulFindMatchingPinsIgnoresCase() {
+    val certificatePinner = CertificatePinner.Builder()
+        .add("EXAMPLE.com", certA1Sha256Pin)
+        .add("*.MyExample.Com", certB1Sha256Pin)
+        .build()
+
+    val expectedPin1 = asList(Pin("EXAMPLE.com", certA1Sha256Pin))
```
Comment (JakeWharton) [inline okhttp/src/test/java/okhttp3/CertificatePinnerKotlinTest.kt]: s/asList/listOf/ here and below

File: okhttp/src/main/java/okhttp3/CertificatePinner.java (hunk 1 above)
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/CertificatePinner.java]: Fixed.

File: okhttp/src/main/java/okhttp3/CertificatePinner.kt (hunk 2 above)
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/CertificatePinner.kt]: Fixed.

File: okhttp/src/main/java/okhttp3/CertificatePinner.kt (hunk 3 above)
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/CertificatePinner.kt]: Fixed.

File: okhttp/src/test/java/okhttp3/CertificatePinnerKotlinTest.kt (hunk 4 above)
Comment (swankjesse) [reply okhttp/src/test/java/okhttp3/CertificatePinnerKotlinTest.kt]: Fixed for all .kt files

File: okhttp-dnsoverhttps/src/main/java/okhttp3/dnsoverhttps/DnsOverHttps.kt [hunk 5]
```diff
@@ -297,7 +296,7 @@ class DnsOverHttps internal constructor(builder: Builder) : Dns {
     }
 
     fun bootstrapDnsHosts(vararg bootstrapDnsHosts: InetAddress): Builder {
-      return bootstrapDnsHosts(asList(*bootstrapDnsHosts))
+      return bootstrapDnsHosts(bootstrapDnsHosts.toList())
```
Comment (yschimke) [inline okhttp-dnsoverhttps/src/main/java/okhttp3/dnsoverhttps/DnsOverHttps.kt]: Thanks

File: okhttp/src/main/java/okhttp3/CertificatePinner.kt [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+  ) {
+    /** The canonical hostname, i.e. `EXAMPLE.com` becomes `example.com`.  */
+    val canonicalHostname: String
+    /** Either `sha1/` or `sha256/`.  */
+    val hashAlgorithm: String
+    /** The hash of the pinned certificate using [.hashAlgorithm].  */
+    val hash: ByteString
+
+    init {
+      this.canonicalHostname = if (pattern.startsWith(WILDCARD)) {
```
Comment (yschimke) [inline okhttp/src/main/java/okhttp3/CertificatePinner.kt]: inline into val above?

File: okhttp/src/main/java/okhttp3/CertificatePinner.kt [hunk 7]
```diff
@@ -0,0 +1,339 @@
...
+        && pins == other.pins
+  }
+
+  override fun hashCode(): Int {
+    var result = Objects.hashCode(certificateChainCleaner)
+    result = 31 * result + pins.hashCode()
+    return result
+  }
+
+  internal class Pin(
```
Comment (yschimke) [inline okhttp/src/main/java/okhttp3/CertificatePinner.kt]: For an internal class is it worth making this a data class?  moving more of the logic into a companion object method?

File: okhttp/src/main/java/okhttp3/CertificatePinner.kt [hunk 8]
```diff
@@ -0,0 +1,339 @@
...
+  fun check(hostname: String, peerCertificates: List<Certificate>) {
+    var peerCertificates = peerCertificates
+    val pins = findMatchingPins(hostname)
+    if (pins.isEmpty()) return
+
+    if (certificateChainCleaner != null) {
+      peerCertificates = certificateChainCleaner.clean(peerCertificates, hostname)
+    }
+
+    for (c in 0 until peerCertificates.size) {
```
Comment (yschimke) [inline okhttp/src/main/java/okhttp3/CertificatePinner.kt]: peerCertificates.foreach?

Comment (yschimke) [reply okhttp/src/main/java/okhttp3/CertificatePinner.kt]: or peerCertificates.find{} ?: throw AssertionError("unsupported hashAlgorithm: ${pin.hashAlgorithm}")
or peerCertificates.any{} || throw AssertionError("unsupported hashAlgorithm: ${pin.hashAlgorithm}")

File: okhttp/src/main/java/okhttp3/CertificatePinner.kt [hunk 9]
```diff
@@ -0,0 +1,339 @@
...
+      val x509Certificate = peerCertificates[c] as X509Certificate
+
+      // Lazily compute the hashes for each certificate.
+      var sha1: ByteString? = null
```
Comment (yschimke) [inline okhttp/src/main/java/okhttp3/CertificatePinner.kt]: TIOLI?   val sha1: ByteString by lazy { sha1(x509Certificate) }

File: okhttp/src/main/java/okhttp3/CertificatePinner.kt (hunk 6 above)
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/CertificatePinner.kt]: I don’t like it for big initializers but I don’t feel strongly.

File: okhttp/src/main/java/okhttp3/CertificatePinner.kt (hunk 7 above)
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/CertificatePinner.kt]: I like it as a data class. For follow-up though.

File: okhttp/src/main/java/okhttp3/CertificatePinner.kt (hunk 8 above)
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/CertificatePinner.kt]: Switched to foreach. No strong feelings about find either way.

File: okhttp/src/main/java/okhttp3/CertificatePinner.kt (hunk 9 above)
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/CertificatePinner.kt]: I like how it reads but I don’t like that it requires an allocation.

Comment (PromanSEW) [reply okhttp/src/main/java/okhttp3/CertificatePinner.kt]: @swankjesse `lazy` does not allocate memory and evaluates only if used

Comment (JakeWharton) [reply okhttp/src/main/java/okhttp3/CertificatePinner.kt]: The instance of the lazy delegate is the allocation

**Pull request merged**